roc/roc.go

```diff
@@ -1,170 +1,142 @@
-	cconfig, _ := config.PassRef()
-	__ret := C.roc_sender_open(ccontext, cconfig)
-	__v := *(**Sender)(unsafe.Pointer(&__ret))
-	return __v
+// safeString ensures that the string is NULL-terminated, a NULL-terminated copy is created otherwise.
+func safeString(str string) string {
+	if len(str) > 0 && str[len(str)-1] != '\x00' {
+		str = str + "\x00"
+	} else if len(str) == 0 {
+		str = "\x00"
+	}
+	return str
 }
 
-// senderBind function as declared in roc/sender.h:134
-func senderBind(sender *Sender, address *Address) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	caddress, _ := address.PassRef()
-	__ret := C.roc_sender_bind(csender, caddress)
-	__v := (int32)(__ret)
-	return __v
-}
+// ============================================================================
+//                              address.h
+// ============================================================================
 
-// senderConnect function as declared in roc/sender.h:153
-func senderConnect(sender *Sender, kind PortType, proto Protocol, address *Address) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	ckind, _ := (C.roc_port_type)(kind), cgoAllocsUnknown
-	cproto, _ := (C.roc_protocol)(proto), cgoAllocsUnknown
-	caddress, _ := address.PassRef()
-	__ret := C.roc_sender_connect(csender, ckind, cproto, caddress)
-	__v := (int32)(__ret)
-	return __v
-}
+// NewAddress parses the `ip`, `port` and `family` and initializes the Address object
+func NewAddress(family Family, ip string, port int) (*Address, error) {
+	a := new(Address)
+	a.mem = make([]byte, C.sizeof_roc_address)
+	a.raw = (*C.roc_address)(unsafe.Pointer(&a.mem))
 
-// senderWrite function as declared in roc/sender.h:177
-func senderWrite(sender *Sender, frame *frame) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	cframe, _ := frame.PassRef()
-	__ret := C.roc_sender_write(csender, cframe)
-	__v := (int32)(__ret)
-	return __v
-}
-
-// senderClose function as declared in roc/sender.h:192
-func senderClose(sender *Sender) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	__ret := C.roc_sender_close(csender)
-	__v := (int32)(__ret)
-	return __v
+	cfamily := (C.roc_family)(family)
+	ip = safeString(ip)
+	cip := C.CString(ip)
+	cport := (C.int)(port)
+	errCode := C.roc_address_init(a.raw, cfamily, cip, cport)
+	runtime.KeepAlive(a.mem)
+
+	if errCode == 0 {
+		return a, nil
+	}
+	if errCode < 0 {
+		return nil, ErrInvalidArguments
+	}
+	return nil, ErrInvalidApi
+}
+
+func (a *Address) Family() (Family, error) {
+	f := C.roc_address_family(a.raw)
+	family := (Family)(f)
+	if family == AfInvalid {
+		return family, ErrInvalidArguments
+	}
+	return family, nil
+}
+
+func (a *Address) IP() (string, error) {
+	const buflen = 255
+	sIP := make([]byte, buflen)
+	res := C.roc_address_ip(a.raw, (*C.char)(unsafe.Pointer(&sIP)), buflen)
```

Comment:
We can allocate sIP on stack instead of the heap. I mean, use array instead of a slice.

Reply:
Hm, no need for this, slice is okay. Go compiler is able to to allocate (small) slices on stack, even created using make: https://go.godbolt.org/z/2dg9Bf (chnage 255 to 255000 to see the difference)

Reply:
awesome!